### ISSUE TITLE
Experimental endpoints (new API)

### ISSUE THREAD

**Issue by druizz90** [wazuh/wazuh#2957]
Hi team,

It is necessary to implement `experimental` endpoints in the new API.

Best regards,

Demetrio.

**Comment by druizz90** [comment]
## Issue update

Below there are examples with `experimental` calls working in the new API:

#### `DELETE experimental/syscheck`

```bash
# curl -X DELETE "http://127.0.0.1:55000/experimental/syscheck"
{
  "data": "Syscheck database deleted",
  "error": 0
}
```

#### `experimental/syscollector/hardware`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/hardware" 
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "board_serial": "unknown",
        "cpu": {
          "cores": 1,
          "mhz": 2712,
          "name": "Intel Core Processor (Broadwell, no TSX, IBRS)"
        },
        "ram": {
          "free": 32760,
          "total": 498828,
          "usage": 94
        },
        "scan": {
          "id": 808862782,
          "time": "2019/03/28 10:54:38"
        }
      }
    ],
    "totalItems": 1
  },
  "error": 0
}
```

#### `/experimental/syscollector/netaddr`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/netaddr"
{
  "data": {
    "items": [
      {
        "address": "192.168.122.111",
        "agent_id": "000",
        "broadcast": "192.168.122.255",
        "iface": "ens3",
        "netmask": "255.255.255.0",
        "proto": "ipv4",
        "scan_id": 1294810199
      },
      {
        "address": "fe80::760a:b8ba:dc8:65a7",
        "agent_id": "000",
        "iface": "ens3",
        "netmask": "ffff:ffff:ffff:ffff::",
        "proto": "ipv6",
        "scan_id": 1294810199
      }
    ],
    "totalItems": 2
  },
  "error": 0
}
```

#### `/experimental/syscollector/netiface`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/netiface"
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "mac": "52:54:00:2b:96:56",
        "mtu": 1500,
        "name": "ens3",
        "rx": {
          "bytes": 7126376,
          "dropped": 0,
          "errors": 0,
          "packets": 14308
        },
        "scan": {
          "id": 1294810199,
          "time": "2019/03/28 10:54:38"
        },
        "state": "up",
        "tx": {
          "bytes": 2899202,
          "dropped": 0,
          "errors": 0,
          "packets": 12626
        },
        "type": "ethernet"
      }
    ],
    "totalItems": 1
  },
  "error": 0
}
```

#### `/experimental/syscollector/netproto`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/netproto"
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "dhcp": "enabled",
        "gateway": "192.168.122.1",
        "iface": "ens3",
        "scan_id": 1294810199,
        "type": "ipv4"
      },
      {
        "agent_id": "000",
        "dhcp": "enabled",
        "iface": "ens3",
        "scan_id": 1294810199,
        "type": "ipv6"
      }
    ],
    "totalItems": 2
  },
  "error": 0
}

```

#### `/experimental/syscollector/os`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/os"     
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "architecture": "x86_64",
        "hostname": "localhost.localdomain",
        "os": {
          "codename": "Core",
          "major": "7",
          "name": "CentOS Linux",
          "platform": "centos",
          "version": "7 (Core)"
        },
        "release": "3.10.0-957.1.3.el7.x86_64",
        "scan": {
          "id": 1783909270,
          "time": "2019/03/28 10:54:38"
        },
        "sysname": "Linux",
        "version": "#1 SMP Thu Nov 29 14:49:43 UTC 2018"
      }
    ],
    "totalItems": 1
  },
  "error": 0
}
```

#### `/experimental/syscollector/packages`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/packages?limit=1"
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "architecture": "x86_64",
        "description": "Utilities from the general purpose cryptography library with TLS implementation",
        "format": "rpm",
        "install_time": "2019/01/14 12:11:20",
        "name": "openssl",
        "scan": {
          "id": 1133540465,
          "time": "2019/03/28 10:54:38"
        },
        "section": "System Environment/Libraries",
        "size": 814,
        "vendor": "CentOS",
        "version": "1:1.0.2k-16.el7"
      }
    ],
    "totalItems": 495
  },
  "error": 0
}
```

#### `/experimental/syscollector/ports`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/ports?limit=2"
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "inode": 466594,
        "local": {
          "ip": "0.0.0.0",
          "port": 1515
        },
        "protocol": "tcp",
        "remote": {
          "ip": "0.0.0.0",
          "port": 0
        },
        "rx_queue": 0,
        "scan": {
          "id": 1399324505,
          "time": "2019/03/28 10:54:47"
        },
        "state": "listening",
        "tx_queue": 0
      },
      {
        "agent_id": "000",
        "inode": 466714,
        "local": {
          "ip": "192.168.122.111",
          "port": 1516
        },
        "protocol": "tcp",
        "remote": {
          "ip": "0.0.0.0",
          "port": 0
        },
        "rx_queue": 0,
        "scan": {
          "id": 1399324505,
          "time": "2019/03/28 10:54:47"
        },
        "state": "listening",
        "tx_queue": 0
      }
    ],
    "totalItems": 9
  },
  "error": 0
}

```

#### `/experimental/syscollector/processes`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/syscollector/processes?limit=1"
{
  "data": {
    "items": [
      {
        "agent_id": "000",
        "argvs": "--switched-root,--system,--deserialize,22",
        "cmd": "/usr/lib/systemd/systemd",
        "egroup": "root",
        "euser": "root",
        "fgroup": "root",
        "name": "systemd",
        "nice": 0,
        "nlwp": 1,
        "pgrp": 1,
        "pid": "1",
        "ppid": 0,
        "priority": 20,
        "processor": 0,
        "resident": 658,
        "rgroup": "root",
        "ruser": "root",
        "scan": {
          "id": 1325104589,
          "time": "2019/03/28 11:01:01"
        },
        "session": 1,
        "sgroup": "root",
        "share": 336,
        "size": 47789,
        "start_time": 4,
        "state": "S",
        "stime": 408,
        "suser": "root",
        "tgid": 1,
        "tty": 0,
        "utime": 103,
        "vm_size": 191156
      }
    ],
    "totalItems": 106
  },
  "error": 0
}
```

#### `/experimental/ciscat/results`

```bash
# curl -X GET "http://127.0.0.1:55000/experimental/ciscat/results"
{
  "data": {
    "items": [],
    "totalItems": 0
  },
  "error": 0
}
```

**Comment by crd1985** [comment]
This issue is blocked by https://github.com/wazuh/wazuh/issues/2843

**Comment by crd1985** [comment]
Issue unblocked

**Comment by druizz90** [comment]
## Issue update

I made `tavern` tests for `experimental` endpoints:

```bash
# /var/ossec/framework/python/bin/pytest -s test_experimental.tavern.yaml -vv
================================================ test session starts =================================================
platform linux -- Python 3.7.2, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /root/wazuh/api
plugins: tavern-0.26.3
collected 10 items                                                                                                   

test_experimental.tavern.yaml::DELETE /experimental/syscheck PASSED
test_experimental.tavern.yaml::GET /experimental/ciscat/results PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/hardware PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/netaddr PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/netiface PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/netproto PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/os PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/packages PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/ports PASSED
test_experimental.tavern.yaml::GET /experimental/syscollector/processes PASSED

================================================== warnings summary ==================================================
/var/ossec/framework/python/lib/python3.7/site-packages/PyYAML-3.13-py3.7-linux-x86_64.egg/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/PyYAML-3.13-py3.7-linux-x86_64.egg/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working

api/test/test_experimental.tavern.yaml::GET /experimental/syscollector/netproto
  /var/ossec/framework/python/lib/python3.7/site-packages/tavern/util/dict_util.py:356: FutureWarning: Expected value was 'null', so this check will pass - this will be removed in a future version. IF you want to check against 'any' value, use '!anything' instead.
    FutureWarning,

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======================================= 10 passed, 2 warnings in 3.04 seconds ========================================

```

**Comment by davidjiglesias** [comment]
I worked on fixing experimental endpoints for new API after fixes in nested fields and merges from 3.10 and 3.11. I also updated and run experimental integration tests and added new environment for automated tests.

### **INTEGRATION TEST RESULTS**
```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.4.0, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python3.6
cachedir: .pytest_cache
rootdir: /home/davidjiglesias/git/wazuh/api
plugins: cov-2.7.1, tavern-0.26.4
collected 10 items                                                                       

test_experimental_endpoints.tavern.yaml::DELETE /experimental/syscheck PASSED      [ 10%]
test_experimental_endpoints.tavern.yaml::GET /experimental/ciscat/results PASSED   [ 20%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/hardware PASSED [ 30%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/netaddr PASSED [ 40%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/netiface PASSED [ 50%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/netproto PASSED [ 60%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/os PASSED  [ 70%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/packages PASSED [ 80%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/ports PASSED [ 90%]
test_experimental_endpoints.tavern.yaml::GET /experimental/syscollector/processes PASSED [100%]

==================================== warnings summary ====================================
test/integration/test_experimental_endpoints.tavern.yaml::DELETE /experimental/syscheck
  /home/davidjiglesias/.local/lib/python3.6/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================= 10 passed, 1 warnings in 4.57 seconds ==========================
```
